controllers/node_controller.go

```diff
@@ -123,8 +136,50 @@ func (r *NodeReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		},
 	}
 
-	return ctrl.NewControllerManagedBy(mgr).
+	err := ctrl.NewControllerManagedBy(mgr).
 		For(&corev1.Node{}).
 		WithEventFilter(onCreationForThisNode).
 		Complete(r)
+	if err != nil {
+		return errors.Wrap(err, "failed to add controller to Node Reconciler listening Node events")
+	}
+
+	// By default all this functors return true so controller watch all events,
+	// but we only want to watch delete/update for current node.
+	onDeleteOrForceUpdateForThisNode := predicate.Funcs{
+		CreateFunc: func(event.CreateEvent) bool {
+			return false
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return nmstate.EventIsForThisNode(deleteEvent.Object)
+		},
+		UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+			return nmstate.EventIsForThisNode(updateEvent.ObjectNew) &&
+				shouldForceRefresh(updateEvent)
+		},
+		GenericFunc: func(event.GenericEvent) bool {
+			return false
+		},
+	}
+	err = ctrl.NewControllerManagedBy(mgr).
+		For(&nmstatev1beta1.NodeNetworkState{}).
+		Watches(&source.Kind{Type: &nmstatev1beta1.NodeNetworkState{}}, &handler.EnqueueRequestForOwner{OwnerType: &corev1.Node{}}, builder.WithPredicates(onDeleteOrForceUpdateForThisNode)).
```

Comment:
I think neither node-controller nor nns-controller use this watch. nns-controller does use `WithEventFilter` though..
can you elaborate what is this change about?

Reply:
This watch listens to events on NodeNetworkState entities. If filters only the NNSs that are removed or the force update label is updated and then passes to the reconcile loop the node that is the owner of this NNS (since the node controller reconcile loops reconciles nodes).
That is very similar to move nns controller was doing just here we need the mapping since the reconcile expect to receive a node and not nns.